File: scrape_alsc.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");
const { chromium } = require("playwright");
function toCsvValue(value) {
  if (value == null) return "";
  const str = String(value).replace(/\r?\n/g, " ");
  if (/[",\n]/.test(str)) {
    return '"' + str.replace(/"/g, '""') + '"';
  }
  return str;
}

function objectsToCsv(rows, headers) {
  const headerLine = headers.map(toCsvValue).join(",");
  const lines = rows.map((row) => headers.map((h) => toCsvValue(row[h])).join(","));
  return [headerLine, ...lines].join("\n");
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function extractBooksFromPage(page) {
  return await page.evaluate(() => {
    const results = [];
    // Each result item appears under an element with class 'result' (based on ALSC directory structure)
    // We defensively search cards that look like search results.
    const items = Array.from(
      document.querySelectorAll(
        ".result, .views-row, .alap-result, .search-result, .node--type-book, .node--view-mode-teaser"
      )
    );

    for (const el of items) {
      const titleEl =
        el.querySelector(".title a, h2 a, h3 a, .node__title a, .field--name-title a") ||
        el.querySelector(".title, h2, h3, .node__title, .field--name-title");
      const authorEl = el.querySelector(
        ".field--name-field-author, .views-field-field-author, .author, .field-author, [class*='author']"
      );
      const publisherEl = el.querySelector(
        ".field--name-field-publisher, .views-field-field-publisher, .publisher, [class*='publisher']"
      );
      const yearEl = el.querySelector(
        ".field--name-field-year, .views-field-field-year, .year, [class*='year']"
      );
      const linkEl = titleEl && titleEl.tagName.toLowerCase() === "a" ? titleEl : el.querySelector("a[href*='/book']");

      const title = titleEl ? titleEl.textContent.trim() : "";
      const author = authorEl ? authorEl.textContent.replace(/^(by\s+)/i, "").trim() : "";
      const publisher = publisherEl ? publisherEl.textContent.trim() : "";
      const yearRaw = yearEl ? yearEl.textContent.trim() : "";
      const yearMatch = yearRaw.match(/\b(19|20)\d{2}\b/);
      const year = yearMatch ? yearMatch[0] : "";
      const url = linkEl ? linkEl.href : "";

      if (title) {
        results.push({ title, author, publisher, year, url });
      }
    }
    return results;
  });
}

async function findAndClickNext(page) {
  // Try different pagination selectors
  const nextSelectors = [
    "a[rel='next']",
    ".pager__item--next a",
    ".pagination a[rel='next']",
    "a:has(span:matches(^Next$|^›$|^»$))",
    "a[aria-label='Next']",
    "a.pager-next, li.pager-next a",
  ];

  for (const sel of nextSelectors) {
    try {
      const has = await page.$(sel);
      if (has) {
        await Promise.all([
          page.waitForLoadState("networkidle", { timeout: 15000 }).catch(() => {}),
          page.click(sel, { timeout: 3000 }).catch(() => {}),
        ]);
        return true;
      }
    } catch (_) {}
  }
  // Heuristic: locate numbered pager and click the active+1 element
  try {
    const clicked = await page.evaluate(() => {
      const pagers = Array.from(document.querySelectorAll(".pager, .pagination, nav[role='navigation']"));
      for (const pager of pagers) {
        const items = Array.from(pager.querySelectorAll("a, span"));
        const activeIndex = items.findIndex((n) => n.classList.contains("is-active") || n.getAttribute("aria-current") === "page");
        if (activeIndex >= 0 && items[activeIndex + 1] && items[activeIndex + 1].tagName.toLowerCase() === "a") {
          (items[activeIndex + 1]).click();
          return true;
        }
      }
      return false;
    });
    if (clicked) {
      await page.waitForLoadState("networkidle", { timeout: 15000 }).catch(() => {});
      return true;
    }
  } catch (_) {}
  return false;
}

async function run() {
  const baseUrl = process.env.ALSC_URL || "https://alsc-awards-shelf.org/directory/results?booklist=14";
  const outJson = path.resolve(process.env.OUT_JSON || "alsc_books.json");
  const outCsv = path.resolve(process.env.OUT_CSV || "alsc_books.csv");
  const maxPages = Number(process.env.MAX_PAGES || 200);

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1280, height: 1600 },
    userAgent:
      "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
  });
  const page = await context.newPage();

  const all = [];

  // Bypass basic bot checks: wait random, scroll, click focus
  await page.goto(baseUrl, { waitUntil: "domcontentloaded", timeout: 30000 });
  await page.waitForLoadState("networkidle", { timeout: 20000 }).catch(() => {});
  await sleep(500 + Math.floor(Math.random() * 1000));
  await page.mouse.move(100 + Math.random() * 600, 200 + Math.random() * 400).catch(() => {});
  await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight / 2)).catch(() => {});

  for (let pageIndex = 1; pageIndex <= maxPages; pageIndex++) {
    // If behind Cloudflare challenge, wait for it to pass up to some time
    const isChallenge = await page.locator("text=Just a moment...").first().isVisible().catch(() => false);
    if (isChallenge) {
      await sleep(4000);
      await page.waitForLoadState("networkidle", { timeout: 10000 }).catch(() => {});
    }

    // Extract on this page
    const items = await extractBooksFromPage(page);
    if (items.length === 0 && isChallenge) {
      // try reload once
      await page.reload({ waitUntil: "domcontentloaded" }).catch(() => {});
      await page.waitForLoadState("networkidle", { timeout: 10000 }).catch(() => {});
      const retryItems = await extractBooksFromPage(page);
      all.push(...retryItems);
    } else {
      all.push(...items);
    }

    // Try to go next; stop if not found
    const moved = await findAndClickNext(page);
    if (!moved) break;
    await sleep(500 + Math.floor(Math.random() * 1200));
  }

  // Deduplicate by title+year
  const dedup = [];
  const seen = new Set();
  for (const r of all) {
    const key = `${r.title}|${r.year}`;
    if (!seen.has(key)) {
      seen.add(key);
      dedup.push(r);
    }
  }

  fs.writeFileSync(outJson, JSON.stringify(dedup, null, 2));
  const csv = objectsToCsv(dedup, ["title", "author", "publisher", "year", "url"]);
  fs.writeFileSync(outCsv, csv);

  console.log(`Saved ${dedup.length} books to:\n- ${outJson}\n- ${outCsv}`);

  await browser.close();
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scrape_alsc.py
import os
import time
import csv
from typing import List, Dict

import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException


def extract_books_from_dom(driver) -> List[Dict[str, str]]:
    results: List[Dict[str, str]] = []
    item_selectors = [
        ".result",
        ".views-row",
        ".alap-result",
        ".search-result",
        ".node--type-book",
        ".node--view-mode-teaser",
    ]
    items = []
    for sel in item_selectors:
        try:
            elements = driver.find_elements(By.CSS_SELECTOR, sel)
            if elements:
                items.extend(elements)
        except Exception:
            pass

    def safe_text(el):
        try:
            return el.text.strip()
        except Exception:
            return ""

    seen = set()
    for el in items:
        title_el = None
        for css in [
            ".title a",
            "h2 a",
            "h3 a",
            ".node__title a",
            ".field--name-title a",
            ".title",
            "h2",
            "h3",
            ".node__title",
            ".field--name-title",
        ]:
            try:
                title_el = el.find_element(By.CSS_SELECTOR, css)
                if title_el:
                    break
            except Exception:
                continue

        author_el = None
        for css in [
            ".field--name-field-author",
            ".views-field-field-author",
            ".author",
            ".field-author",
            "[class*='author']",
        ]:
            try:
                author_el = el.find_element(By.CSS_SELECTOR, css)
                if author_el:
                    break
            except Exception:
                continue

        publisher_el = None
        for css in [
            ".field--name-field-publisher",
            ".views-field-field-publisher",
            ".publisher",
            "[class*='publisher']",
        ]:
            try:
                publisher_el = el.find_element(By.CSS_SELECTOR, css)
                if publisher_el:
                    break
            except Exception:
                continue

        year_el = None
        for css in [
            ".field--name-field-year",
            ".views-field-field-year",
            ".year",
            "[class*='year']",
        ]:
            try:
                year_el = el.find_element(By.CSS_SELECTOR, css)
                if year_el:
                    break
            except Exception:
                continue

        link_el = None
        try:
            if title_el and title_el.tag_name.lower() == "a":
                link_el = title_el
        except Exception:
            link_el = None
        if link_el is None:
            try:
                link_el = el.find_element(By.CSS_SELECTOR, "a[href*='/book']")
            except Exception:
                link_el = None

        title = safe_text(title_el) if title_el else ""
        author = safe_text(author_el).lstrip("by ") if author_el else ""
        publisher = safe_text(publisher_el) if publisher_el else ""
        year_text = safe_text(year_el) if year_el else ""
        year = ""
        for token in year_text.split():
            if token.isdigit() and len(token) == 4 and token.startswith(("19", "20")):
                year = token
                break

        url = ""
        try:
            url = link_el.get_attribute("href") if link_el else ""
        except Exception:
            url = ""

        if title:
            key = f"{title}|{year}"
            if key not in seen:
                seen.add(key)
                results.append({
                    "title": title,
                    "author": author,
                    "publisher": publisher,
                    "year": year,
                    "url": url,
                })
    return results


def click_next(driver) -> bool:
    selectors = [
        "a[rel='next']",
        ".pager__item--next a",
        ".pagination a[rel='next']",
        "a[aria-label='Next']",
        "a.pager-next, li.pager-next a",
    ]
    for sel in selectors:
        try:
            el = driver.find_element(By.CSS_SELECTOR, sel)
            if el:
                el.click()
                return True
        except Exception:
            continue
    # heuristic: active + 1
    try:
        pagers = driver.find_elements(By.CSS_SELECTOR, ".pager, .pagination, nav[role='navigation']")
        for pager in pagers:
            anchors = pager.find_elements(By.CSS_SELECTOR, "a, span")
            active_index = -1
            for i, n in enumerate(anchors):
                try:
                    if "is-active" in n.get_attribute("class" or "") or (n.get_attribute("aria-current") == "page"):
                        active_index = i
                        break
                except Exception:
                    pass
            if active_index >= 0 and active_index + 1 < len(anchors):
                if anchors[active_index + 1].tag_name.lower() == "a":
                    anchors[active_index + 1].click()
                    return True
    except Exception:
        pass
    return False


def main():
    base_url = os.environ.get("ALSC_URL", "https://alsc-awards-shelf.org/directory/results?booklist=14")
    out_json = os.environ.get("OUT_JSON", "alsc_books.json")
    out_csv = os.environ.get("OUT_CSV", "alsc_books.csv")
    max_pages = int(os.environ.get("MAX_PAGES", "200"))

    options = uc.ChromeOptions()
    options.add_argument("--headless=new")
    options.add_argument("--no-sandbox")
    options.add_argument("--disable-gpu")
    options.add_argument("--window-size=1280,1600")

    driver = uc.Chrome(options=options)
    driver.set_page_load_timeout(60)

    driver.get(base_url)
    time.sleep(3)

    all_rows: List[Dict[str, str]] = []
    for page_index in range(1, max_pages + 1):
        try:
            WebDriverWait(driver, 15).until(EC.presence_of_element_located((By.TAG_NAME, "body")))
        except TimeoutException:
            pass

        if "Just a moment" in driver.page_source:
            time.sleep(5)

        rows = extract_books_from_dom(driver)
        if not rows and "Just a moment" in driver.page_source:
            driver.refresh()
            time.sleep(3)
            rows = extract_books_from_dom(driver)
        all_rows.extend(rows)

        moved = click_next(driver)
        if not moved:
            break
        time.sleep(1.2)

    # dedupe
    deduped = []
    seen = set()
    for r in all_rows:
        key = f"{r['title']}|{r['year']}"
        if key not in seen:
            seen.add(key)
            deduped.append(r)

    # write JSON
    import json
    with open(out_json, "w", encoding="utf-8") as f:
        json.dump(deduped, f, ensure_ascii=False, indent=2)

    # write CSV
    headers = ["title", "author", "publisher", "year", "url"]
    with open(out_csv, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=headers)
        writer.writeheader()
        for r in deduped:
            writer.writerow(r)

    print(f"Saved {len(deduped)} books to: {out_json}, {out_csv}")

    driver.quit()


if __name__ == "__main__":
    main()

